bolt-sidecar/src/api/commitments/firewall_recv.rs

```diff
@@ -0,0 +1,622 @@
+use futures::stream::{FuturesUnordered, SplitSink, SplitStream};
+use futures::{FutureExt, SinkExt, StreamExt};
+use std::collections::VecDeque;
+use std::fmt::Debug;
+use std::fmt::{self, Formatter};
+use std::task::Poll;
+use std::time::Duration;
+use std::{future::Future, pin::Pin};
+use tokio::net::TcpStream;
+use tokio::sync::broadcast::error::TryRecvError;
+use tokio::sync::{broadcast, mpsc, oneshot};
+use tokio_tungstenite::tungstenite::protocol::WebSocketConfig;
+use tokio_tungstenite::tungstenite::Message;
+use tokio_tungstenite::{connect_async_with_config, MaybeTlsStream, WebSocketStream};
+use tracing::{error, info, trace, warn};
+
+use reqwest::Url;
+
+use crate::common::backoff::retry_with_backoff;
+use crate::common::secrets::EcdsaSecretKeyWrapper;
+use crate::config::chain::Chain;
+use crate::primitives::commitment::SignedCommitment;
+use crate::primitives::{CommitmentRequest, InclusionRequest};
+
+use super::server::CommitmentEvent;
+use super::spec::CommitmentError;
+
+/// The interval at which to send ping messages from connected clients.
+#[cfg(test)]
+const PING_INTERVAL: Duration = Duration::from_secs(3);
+#[cfg(not(test))]
+const PING_INTERVAL: Duration = Duration::from_secs(30);
+
+/// The maximum number of retries to attempt when reconnecting to a websocket server.
+const MAX_RETRIES: usize = 10;
+
+type ShutdownSignal = Pin<Box<dyn Future<Output = ()> + Send>>;
```

Comment:
Can also use `CancellationToken`s here: https://tokio.rs/tokio/topics/shutdown. Would actually be a good pattern to implement on all async tasks for graceful shutdown.

Reply:
This is prob overkill, but remember seeing it in the wild and it was cool :) 
https://github.com/superfly/corrosion/blob/edf04b0bc9bc6e2ae0dd42d69cb4368a2976eb24/crates/tripwire/src/tripwire.rs#L21